sourcegraph: multi-tenant Zoekt

stefanhengl opened the pull request:
This updates webserver and sourcegraph-indexserver to support multi-tenancy.

The change is behind an ENV feature-flag.

Key changes:
- tenant ID is now part of the index (repo metadata)
- GRPC: IndexOption and Repository have a new field `TenantId`
- If multi-tenancy is enabled, webserver checks if tenant in context matches the tenant id in the shard
- zoekt-git-index has a new parameter "-shard_prefix <string>". If set, the value will be used instead of repository name as prefix for the name of the shard. For Sourcegraph we use "\<tenant id\>_\<repository id\>" as prefix if multi-tenancy is enabled (see screenshot).

Assumptions:
  - All calls to Sourcegraph are privileged 

Test plan:
- New tests
- Ran this together with Sourcegraph (with and without MT enabled)

This is how the new Sourcegraph naming scheme looks like on disk:
<img width="842" alt="image" src="https://github.com/user-attachments/assets/173c3f52-93be-4401-a552-7e7682f5f444">

stefanhengl committed:
add tenant id to shard

stefanhengl committed:
use Sourcegraph ID for shard names

stefanhengl committed:
support tenants

stefanhengl committed:
update tests

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/index.go:
```diff
@@ -117,6 +132,7 @@ func (o *indexArgs) BuildOptions() *build.Options {
 				"archived": marshalBool(o.Archived),
 				// Calculate repo rank based on the latest commit date.
 				"latestCommitDate": "1",
+				"tenantid":         strconv.Itoa(tenantID),
```
very very minor, to match the other name styles this should be `tenantID`

stefanhengl replied:
I am matching styles with repoid 7 lines up ;-)

keegancsmith reviewed (commented):
This feels so much better than the other PR right? dig it.

stefanhengl committed:
PR comments

- prefix flag instead of bool
- padded ids
- include tenant in id
- add tenant tests

stefanhengl committed:
webserver.proto: add TenantId to Repository

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/index.go:
```diff
@@ -95,17 +98,26 @@ type indexArgs struct {
 func (o *indexArgs) BuildOptions() *build.Options {
+	shardPrefix := ""
+	if o.IdBasedNames {
+		shardPrefix = fmt.Sprintf("%09d_%09d", o.TenantID, o.IndexOptions.RepoID)
```
I like it

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/index_test.go:
```diff
@@ -603,6 +610,7 @@ func TestIndex(t *testing.T) {
 			"git -C $TMPDIR/test%2Frepo.git config zoekt.priority 0",
 			"git -C $TMPDIR/test%2Frepo.git config zoekt.public 0",
 			"git -C $TMPDIR/test%2Frepo.git config zoekt.repoid 0",
+			"git -C $TMPDIR/test%2Frepo.git config zoekt.tenantID 1",
```
we need a test for when we set -shard_prefix

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -592,6 +601,7 @@ func (s *Server) Index(args *indexArgs) (state indexState, err error) {
 		case build.IndexStateMeta:
 			log.Printf("updating index.meta %s", args.String())
 
+			// TODO(stefan) handle mergeMeta for tenant id.
```
:O also need exploding then to understand the new naming scheme?

stefanhengl replied:
This here is a bit different. We should probably trigger a reindex if the tenant id changes, but I didn't want to include it now because that would trigger a reindex for everyone.

For explode, I updated the naming gated by tenant enforcement. I will file a follow-up to make that nicer. It's obviously too brittle.

jtibshirani commented on build/e2e_test.go:
```diff
@@ -153,6 +158,88 @@ func TestBasic(t *testing.T) {
+		t.Fatalf("Parse(111): %v", err)
+	}
+
+	var sOpts zoekt.SearchOptions
```
It'd be great to have end-to-end tests for `List` too. I am totally paranoid about this sort of thing to sleep better at night.

stefanhengl committed:
Update build/builder.go

Co-authored-by: Keegan Carruthers-Smith <keegan.csmith@gmail.com>

stefanhengl committed:
Update build/builder.go

Co-authored-by: Keegan Carruthers-Smith <keegan.csmith@gmail.com>

stefanhengl commented on cmd/zoekt-sourcegraph-indexserver/index.go:
```diff
@@ -95,17 +98,26 @@ type indexArgs struct {
 	ShardMerging bool
+
+	// IdBasedNames is true if we want to use ID-based names as prefix for the shard name.
+	IdBasedNames bool
```
I like the first one better, it seems easier to read. golang/go seems to use both https://sourcegraph.sourcegraph.com/search?q=context:global+r:golang/go+Id&patternType=regexp&case=yes&sm=0.

stefanhengl committed:
more tests, rename ENV

stefanhengl committed:
remove ENV, gate on EnforceTenant

stefanhengl committed:
teach explode about id-based shards

stefanhengl committed:
fix typo

stefanhengl committed:
add SECURITY notices
